hbase-server/src/main/java/org/apache/hadoop/hbase/master/procedure/SnapshotProcedure.java

```diff
@@ -0,0 +1,422 @@
+    try {
+      ModifyRegionUtils.editRegions(exec, regions, new ModifyRegionUtils.RegionEditTask() {
+        @Override
+        public void editRegion(final RegionInfo regionInfo) throws IOException {
+          snapshotManifest.addRegion(CommonFSUtils.getTableDir(rootDir, snapshotTable), regionInfo);
+        }
+      });
+    } finally {
+      exec.shutdown();
+    }
+    status.setStatus("Completed referencing offline regions of table: " + snapshotTable);
+  }
+
+  private void snapshotMobRegion() throws IOException {
+    RegionInfo mobRegionInfo = MobUtils.getMobRegionInfo(htd.getTableName());
+    snapshotManifest.addMobRegion(mobRegionInfo);
+    status.setStatus("Completed referencing HFiles for the mob region of table: " + snapshotTable);
+  }
+
+  @Override
+  protected void serializeStateData(ProcedureStateSerializer serializer) throws IOException {
+    super.serializeStateData(serializer);
+    serializer.serialize(SnapshotProcedureStateData
+      .newBuilder().setSnapshot(this.snapshot).build());
+  }
+
+  @Override
+  protected void deserializeStateData(ProcedureStateSerializer serializer) throws IOException {
+    super.deserializeStateData(serializer);
+    SnapshotProcedureStateData data = serializer.deserialize(SnapshotProcedureStateData.class);
+    this.snapshot = data.getSnapshot();
+  }
+
+  private Procedure<MasterProcedureEnv>[] createRemoteSnapshotProcedures(MasterProcedureEnv env) {
+    return env.getAssignmentManager().getTableRegions(snapshotTable, true)
+      .stream().filter(r -> RegionReplicaUtil.isDefaultReplica(r))
+      .map(r -> new SnapshotRegionProcedure(snapshot, r))
+      .toArray(SnapshotRegionProcedure[]::new);
+  }
+
+  // here we assign region snapshot manifest verify tasks to all region servers
+  // in cluster with the help of master load balancer.
+  private Procedure<MasterProcedureEnv>[] createRemoteVerifyProcedures(MasterProcedureEnv env)
+    throws IOException {
+    List<RegionInfo> regions = env
+      .getAssignmentManager().getTableRegions(snapshotTable, false);
+    List<ServerName> servers = env
+      .getMasterServices().getServerManager().getOnlineServersList();
+    return env.getMasterServices().getLoadBalancer()
+      .roundRobinAssignment(regions, servers).entrySet().stream()
+      .map(e -> new SnapshotVerifyProcedure(snapshot, e.getValue(), e.getKey(), regions.size()))
+      .toArray(SnapshotVerifyProcedure[]::new);
+  }
+
+  @Override
+  public void toStringClassDetails(StringBuilder builder) {
+    builder.append(getClass().getName())
+      .append(", id=").append(getProcId())
+      .append(", snapshot=").append(ClientSnapshotDescriptionUtils.toString(snapshot));
+  }
+
+  public SnapshotDescription getSnapshotDesc() {
+    return snapshot;
+  }
+
+  @Override
+  protected void afterReplay(MasterProcedureEnv env) {
+    try {
+      prepareSnapshotEnv(env);
+    } catch (IOException e) {
+      LOG.error("Failed replaying {}, mark procedure as failed", this, e);
```

Comment:
Is this possible?

Reply:
Haven't encountered it so far. But we may need to be prepared for filesystem failure.

Reply:
OK, here you just call prepareSnapshotEnv... I do not think this is correct, in afterReplay typically you should not do the same thing with a normal step? I was thinking we will use this method to restore the in progress snapshot procedures map in SnapshotManager...

Reply:
My idea is like this, before master restarts, if the snapshot is already running, then there will be no split/merge procedure running, if the snapshot is not running, then we will start in the initial state, and we will check if there are any split/merge procedure running in this state. I have not thought of a situation where this would happen. so I just call prepareSnapshotEnv. But I agree with that it is more safe to check again, I will fix this. Thanks duo @Apache9

Reply:
There are some variables such as workingDir, workingDirFs or snapshotManifest, they are runtime variables and maybe we don't need to persist them to the procedure store, just reconstruct them after replay. Based on this consideration, I call the prepareSnapshotEnv in this method.

The SnapshotManager has not been initialized when we replay procedures. As we discussed before, the solution is changing the startup order of master or using a map to track the unfinished snapshot procedure and pass the map when initializing SnapshotManager. I am trying the second solution and have not finished this work. The code is a little ugly, because I have to put some special logic to the procedure-v2 framework which is what I want to avoid. So I am thinking about the first solution too ....